dali/python/nvidia/dali/external_source.pyi

```diff
@@ -0,0 +1,115 @@
+# Copyright (c) 2020-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Union, Optional, overload
+from typing import Sequence, Any, Callable, Iterable
+
+from nvidia.dali.data_node import DataNode
+from nvidia.dali.types import DALIDataType, DALIImageType, DALIInterpType
+
+
+class ExternalSource:
+
+    def __init__(
+        self,
+        source: Optional[Union[Callable[..., Any], Iterable[Any], Callable[..., Sequence[Any]], Iterable[Sequence[Any]]]] = None,
```

Comment:
A random thought, but should we go into details like `Sequence[Any]` of the source return type when we already list Any as acceptable type?

Reply:
This is a really good point, my thinking is basically `Any` ~ `Tensor` and `Sequence[Any]` ~ batch of tensors.
The problem is I don't have any (heh) good approximation of what Tensor can be, there is a proposal to have something like this in the newer Pythons https://github.com/python/typeshed/blob/2a0fc1b582ef84f7a82c0beb39fa617de2539d3d/stdlib/_typeshed/__init__.pyi#L194
but is still doesn't encapsulate GPU tensors.
I want to express the outermost kind of type, and we can iteratively get better, I am not yet fully sure what is going on here: https://github.com/numpy/numpy/blob/main/numpy/_typing/_array_like.py

Reply:
I assumed that's the intention and it feels like a good idea. My concern is whether it's best to mention the Sequence now or make it more detailed once we have some TensorLike type. For example, replacing Any with the TensorLike, probably won't be enough. I think we also need Tuple[Sequence[TensorLike]] and Tuple[TensorLike] for the sample mode.

Reply:
Oh, sorry, I forgot my intention here. So the Sequence comes from the multiple outputs, looking at the `external_source` overloads, but I didn't want to go too much into the overloads of the class API. 
Looks like we can't put `num_outputs` in the call, and the `source` is kw-only there, will fix.

The type for source is the union of two overloads I did for the fn API, basically `Any` can be either TensorLike or TensorBatchLike. `TensorLike` can be used in sample mode, and `Union[TensorLike, TensorBatchLike]` can be used in batch mode. And than we have Sequence of those in case of `num_outputs != None`. It is a lot of options.